pkg/sfu/buffer/dependencydescriptorparser.go

```diff
@@ -12,91 +13,143 @@ import (
+		var maxSpatial, maxTemporal int32
+		for _, dt := range r.decodeTargets {
+			if *mask&(1<<dt.Target) != uint32(dd.DecodeTargetNotPresent) {
+				if maxSpatial < dt.Layer.Spatial {
+					maxSpatial = dt.Layer.Spatial
 				}
-				if layer.Temporal > maxTemporal {
-					maxTemporal = layer.Temporal
+				if maxTemporal < dt.Layer.Temporal {
+					maxTemporal = dt.Layer.Temporal
 				}
-				r.decodeTargetLayer = append(r.decodeTargetLayer, layer)
 			}
-			r.logger.Debugw("max layer changed", "maxSpatial", maxSpatial, "maxTemporal", maxTemporal)
-			go r.onMaxLayerChanged(maxSpatial, maxTemporal)
 		}
+		r.logger.Debugw("max layer changed", "maxSpatial", maxSpatial, "maxTemporal", maxTemporal)
+		// DD-TODO: should this be called in a goroutine as the preivous invocation above is?
+		r.onMaxLayerChanged(maxSpatial, maxTemporal)
+	}
 
-		if ddVal.AttachedStructure != nil && ddVal.FirstPacketInFrame {
-			r.logger.Debugw(fmt.Sprintf("parsed dependency descriptor\n%s", ddVal.String()))
-		}
+	withDecodeTargets := &DependencyDescriptorWithDecodeTarget{
+		Descriptor:    &ddVal,
+		DecodeTargets: r.decodeTargets,
+	}
+
+	return withDecodeTargets, videoLayer, nil
+}
 
-		if mask := ddVal.ActiveDecodeTargetsBitmask; mask != nil {
-			var maxSpatial, maxTemporal int32
-			for dt, layer := range r.decodeTargetLayer {
-				if *mask&(1<<dt) != uint32(dd.DecodeTargetNotPresent) {
-					if maxSpatial < layer.Spatial {
-						maxSpatial = layer.Spatial
-					}
-					if maxTemporal < layer.Temporal {
-						maxTemporal = layer.Temporal
-					}
+// ------------------------------------------------------------------------------
+
+type DependencyDescriptorDecodeTarget struct {
+	Target int
+	Layer  VideoLayer
+}
+
+func ProcessFrameDependencyStructure(structure *dd.FrameDependencyStructure) []DependencyDescriptorDecodeTarget {
+	decodeTargets := make([]DependencyDescriptorDecodeTarget, 0, structure.NumDecodeTargets)
+	for target := 0; target < structure.NumDecodeTargets; target++ {
+		layer := VideoLayer{Spatial: 0, Temporal: 0}
+		for _, t := range structure.Templates {
+			if t.DecodeTargetIndications[target] != dd.DecodeTargetNotPresent {
+				if layer.Spatial < int32(t.SpatialId) {
+					layer.Spatial = int32(t.SpatialId)
+				}
+				if layer.Temporal < int32(t.TemporalId) {
+					layer.Temporal = int32(t.TemporalId)
 				}
 			}
-			r.logger.Debugw("max layer changed", "maxSpatial", maxSpatial, "maxTemporal", maxTemporal)
-			r.onMaxLayerChanged(maxSpatial, maxTemporal)
 		}
-		return &ddVal, videoLayer, nil
+		decodeTargets = append(decodeTargets, DependencyDescriptorDecodeTarget{target, layer})
+	}
+
+	// sort decode target layer by spatial and temporal from high to low
+	sort.Slice(decodeTargets, func(i, j int) bool {
+		return decodeTargets[i].Layer.GreaterThan(decodeTargets[j].Layer)
+	})
+
+	return decodeTargets
+}
+
+func GetActiveDecodeTargetBitmask(layer VideoLayer, decodeTargets []DependencyDescriptorDecodeTarget) *uint32 {
```

Comment:
This is called from down track side based on its target layer. Pulling this in here as a stateless function to accumulate knowledge of DD closer together. Actually, I wanted to put these in `pkg/sfu/dependencydescriptor` directory, but could not because these functions use `VideoLayer` struct and that causes a circular import :-(